hspec-core/src/Test/Hspec/Core/Spec/Vault.hs

```diff
@@ -0,0 +1,27 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Test.Hspec.Core.Spec.Vault (
+  Vault
+, setValue
+, getValue
+) where
+
+import           Prelude ()
+import           Test.Hspec.Core.Compat
+
+import           Data.Typeable
+import           Data.Dynamic
+
+newtype Vault = Vault [(TypeRep, Dynamic)]
+  deriving (
+#if MIN_VERSION_base(4,11,0)
+    Semigroup,
+#endif
+    Monoid)
+
+setValue :: Typeable v => v -> Vault -> Vault
+setValue v (Vault vs) = Vault $ (typeOf v, toDyn v) : vs
+
+getValue :: forall v. Typeable v => Vault -> Maybe v
+getValue (Vault vs) = lookup (typeOf (undefined :: v)) vs >>= fromDynamic
```

Comment:
I imagine we can depend on `containers`?

```suggestion
newtype Vault = Vault (Map TypeRep Dynamic)
  deriving (
#if MIN_VERSION_base(4,11,0)
    Semigroup,
#endif
    Monoid)

setValue :: Typeable v => v -> Vault -> Vault
setValue v (Vault vs) = Vault $ Map.insert (typeOf v) (toDyn v) vs

getValue :: forall v. Typeable v => Vault -> Maybe v
getValue (Vault vs) = Map.lookup (typeOf (undefined :: v)) vs >>= fromDynamic
```

Reply:
Heya 👋

Let's wait until we know if we are going to keep this code.

The short story is:
1. In the common case of storing values of a single type, you are only ever looking at the head of the list.
2. Even if you are going to store values of different types, I think you would need **a lot of nesting** to be able to observe a difference between map/list.
3. We are not dealing with untrusted user input; still, everything else equal, logarithmic is of course better than linear.